dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,337 @@
+            sbStore.append(leftBracket).append("then=");
+            sbStore.append(visualizeStore(thenStore));
+            S elseStore = input.getElseStore();
+            sbStore.append(", else=");
+            sbStore.append(visualizeStore(elseStore));
+            sbStore.append(rightBracket);
+        }
+        // separator
+        sbStore.append(escapeCharacter).append("~~~~~~~~~").append(escapeCharacter);
+
+        return sbStore.toString();
+    }
+
+    protected String visualizeSpecialBlockHelper(SpecialBlock sbb, String separator) {
+        String specialBlock = "";
+        switch (sbb.getSpecialType()) {
+            case ENTRY:
+                specialBlock = "<entry>" + separator;
+                break;
+            case EXIT:
+                specialBlock = "<exit>" + separator;
+                break;
+            case EXCEPTIONAL_EXIT:
+                specialBlock = "<exceptional-exit>" + separator;
+                break;
+        }
+        return specialBlock;
+    }
+
+    protected Node getLastNode(Block bb) {
+        Node lastNode;
+        switch (bb.getType()) {
+            case REGULAR_BLOCK:
+                List<Node> blockContents = ((RegularBlock) bb).getContents();
+                lastNode = blockContents.get(blockContents.size() - 1);
+                break;
+            case EXCEPTION_BLOCK:
+                lastNode = ((ExceptionBlock) bb).getNode();
+                break;
+            default:
+                lastNode = null;
+        }
+        return lastNode;
+    }
+
+    @Override
+    public String visualizeBlockNode(Node t, @Nullable Analysis<A, S, T> analysis) {
+        StringBuilder sbBlockNode = new StringBuilder();
+        sbBlockNode
+                .append(prepareString(t.toString()))
+                .append("   [ ")
+                .append(prepareNodeType(t))
+                .append(" ]");
+        if (analysis != null) {
+            A value = analysis.getValue(t);
+            if (value != null) {
+                sbBlockNode.append("    > ").append(prepareString(value.toString()));
+            }
+        }
+        return sbBlockNode.toString();
+    }
+
+    /**
+     * Generate the order of processing blocks.
+     *
+     * @param cfg the current control flow graph
+     */
+    IdentityHashMap<Block, List<Integer>> getProcessOrder(ControlFlowGraph cfg) {
+        IdentityHashMap<Block, List<Integer>> depthFirstOrder = new IdentityHashMap<>();
+        int count = 1;
+        for (Block b : cfg.getDepthFirstOrderedBlocks()) {
+            depthFirstOrder.computeIfAbsent(b, k -> new ArrayList<>());
+            depthFirstOrder.get(b).add(count++);
+        }
+        return depthFirstOrder;
+    }
+
+    @Override
+    public String visualizeStore(S store) {
+        return store.visualize(this);
+    }
+
+    @Override
+    public String visualizeStoreFooter() {
+        return ")";
+    }
+
+    protected String prepareNodeType(Node t) {
```

Comment:
For both `prepareXXX` methods, can you find more descriptive names?

Reply:
What about `prepareNodeType` => `simplifyNodeType`?

Reply:
How about `getNodeSimpleName`? Both `prepare` and `simplify` sound like they are doing more...

Reply:
Sounds good, I will do it in this way.